File: crates/solobase/src/blocks/legalpages.rs
use std::collections::HashMap;
use wafer_run::block::{Block, BlockInfo, AdminUIInfo};
use wafer_run::context::Context;
use wafer_run::types::*;
use wafer_run::helpers::*;
use wafer_core::clients::database as db;
use wafer_core::clients::database::{Filter, FilterOp, ListOptions, SortField};

pub struct LegalPagesBlock;

const COLLECTION: &str = "ext_legalpages_legal_documents";

impl LegalPagesBlock {
    fn handle_get_public(&self, ctx: &dyn Context, msg: &mut Message, doc_type: &str) -> Result_ {
        // Find published document of given type
        let opts = ListOptions {
            filters: vec![
                Filter {
                    field: "doc_type".to_string(),
                    operator: FilterOp::Equal,
                    value: serde_json::Value::String(doc_type.to_string()),
                },
                Filter {
                    field: "status".to_string(),
                    operator: FilterOp::Equal,
                    value: serde_json::Value::String("published".to_string()),
                },
            ],
            sort: vec![SortField { field: "version".to_string(), desc: true }],
            limit: 1,
            ..Default::default()
        };

        let result = match db::list(ctx, COLLECTION, &opts) {
            Ok(r) => r,
            Err(e) => return err_internal(msg, &format!("Database error: {e}")),
        };

        if result.records.is_empty() {
            let html = format!(
                "<html><body><h1>{}</h1><p>No {} document has been published yet.</p></body></html>",
                if doc_type == "terms" { "Terms of Service" } else { "Privacy Policy" },
                doc_type
            );
            return respond(msg, html.into_bytes(), "text/html; charset=utf-8");
        }

        let record = &result.records[0];
        let raw_content = record.data.get("content").and_then(|v| v.as_str()).unwrap_or("");
        let content = sanitize_html(raw_content);
        let title = record.data.get("title").and_then(|v| v.as_str()).unwrap_or(doc_type)
            .replace('&', "&amp;").replace('<', "&lt;").replace('>', "&gt;");

        let html = format!(
            r#"<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>{title}</title>
<style>body{{font-family:system-ui,sans-serif;max-width:800px;margin:40px auto;padding:0 20px;line-height:1.6;color:#333}}h1{{color:#111}}</style>
</head><body><h1>{title}</h1><div>{content}</div></body></html>"#
        );
        respond(msg, html.into_bytes(), "text/html; charset=utf-8")
    }

    fn handle_admin_ui(&self, msg: &mut Message) -> Result_ {
        respond(msg, ADMIN_HTML.as_bytes().to_vec(), "text/html; charset=utf-8")
    }

    fn handle_admin_list(&self, ctx: &dyn Context, msg: &mut Message) -> Result_ {
        let (_, page_size, offset) = msg.pagination_params(20);
        let doc_type = msg.query("type");
        let mut filters = Vec::new();
        if !doc_type.is_empty() {
            filters.push(Filter {
                field: "doc_type".to_string(),
                operator: FilterOp::Equal,
                value: serde_json::Value::String(doc_type.to_string()),
            });
        }
        let opts = ListOptions {
            filters,
            sort: vec![SortField { field: "updated_at".to_string(), desc: true }],
            limit: page_size as i64,
            offset: offset as i64,
        };
        match db::list(ctx, COLLECTION, &opts) {
            Ok(result) => json_respond(msg, &result),
            Err(e) => err_internal(msg, &format!("Database error: {e}")),
        }
    }

    fn handle_admin_get(&self, ctx: &dyn Context, msg: &mut Message) -> Result_ {
        let id = msg.var("id");
        if id.is_empty() {
            return err_bad_request(msg, "Missing document ID");
        }
        match db::get(ctx, COLLECTION, id) {
            Ok(record) => json_respond(msg, &record),
            Err(e) if e.code == "not_found" => err_not_found(msg, "Document not found"),
            Err(e) => err_internal(msg, &format!("Database error: {e}")),
        }
    }

    fn handle_admin_create(&self, ctx: &dyn Context, msg: &mut Message) -> Result_ {
        #[derive(serde::Deserialize)]
        struct CreateDoc {
            doc_type: String,
            title: String,
            content: String,
        }
        let body: CreateDoc = match msg.decode() {
            Ok(b) => b,
            Err(e) => return err_bad_request(msg, &format!("Invalid body: {e}")),
        };

        let now = chrono::Utc::now().to_rfc3339();
        let mut data = HashMap::new();
        data.insert("doc_type".to_string(), serde_json::Value::String(body.doc_type));
        data.insert("title".to_string(), serde_json::Value::String(body.title));
        data.insert("content".to_string(), serde_json::Value::String(body.content));
        data.insert("status".to_string(), serde_json::Value::String("draft".to_string()));
        data.insert("version".to_string(), serde_json::Value::Number(1.into()));
        data.insert("created_at".to_string(), serde_json::Value::String(now.clone()));
        data.insert("updated_at".to_string(), serde_json::Value::String(now));
        data.insert("created_by".to_string(), serde_json::Value::String(msg.user_id().to_string()));

        match db::create(ctx, COLLECTION, data) {
            Ok(record) => json_respond(msg, &record),
            Err(e) => err_internal(msg, &format!("Database error: {e}")),
        }
    }

    fn handle_admin_update(&self, ctx: &dyn Context, msg: &mut Message) -> Result_ {
        let id = msg.var("id");
        if id.is_empty() {
            return err_bad_request(msg, "Missing document ID");
        }

        let body: HashMap<String, serde_json::Value> = match msg.decode() {
            Ok(b) => b,
            Err(e) => return err_bad_request(msg, &format!("Invalid body: {e}")),
        };

        let mut data = body;
        data.insert("updated_at".to_string(), serde_json::Value::String(chrono::Utc::now().to_rfc3339()));

        match db::update(ctx, COLLECTION, id, data) {
            Ok(record) => json_respond(msg, &record),
            Err(e) if e.code == "not_found" => err_not_found(msg, "Document not found"),
            Err(e) => err_internal(msg, &format!("Database error: {e}")),
        }
    }

    fn handle_admin_publish(&self, ctx: &dyn Context, msg: &mut Message) -> Result_ {
        let id = msg.var("id");
        if id.is_empty() {
            return err_bad_request(msg, "Missing document ID");
        }

        // Get current document
        let doc = match db::get(ctx, COLLECTION, id) {
            Ok(r) => r,
            Err(e) if e.code == "not_found" => return err_not_found(msg, "Document not found"),
            Err(e) => return err_internal(msg, &format!("Database error: {e}")),
        };

        let doc_type = doc.data.get("doc_type").and_then(|v| v.as_str()).unwrap_or("").to_string();

        // Unpublish other documents of same type
        let existing = db::list_all(
            ctx,
            COLLECTION,
            vec![
                Filter { field: "doc_type".to_string(), operator: FilterOp::Equal, value: serde_json::Value::String(doc_type) },
                Filter { field: "status".to_string(), operator: FilterOp::Equal, value: serde_json::Value::String("published".to_string()) },
            ],
        );
        if let Ok(records) = existing {
            for r in records {
                let mut upd = HashMap::new();
                upd.insert("status".to_string(), serde_json::Value::String("archived".to_string()));
                if let Err(e) = db::update(ctx, COLLECTION, &r.id, upd) {
                    tracing::warn!("Failed to archive previous legal page version: {e}");
                }
            }
        }

        // Publish this one
        let mut data = HashMap::new();
        data.insert("status".to_string(), serde_json::Value::String("published".to_string()));
        data.insert("published_at".to_string(), serde_json::Value::String(chrono::Utc::now().to_rfc3339()));
        data.insert("updated_at".to_string(), serde_json::Value::String(chrono::Utc::now().to_rfc3339()));

        match db::update(ctx, COLLECTION, id, data) {
            Ok(record) => json_respond(msg, &record),
            Err(e) => err_internal(msg, &format!("Database error: {e}")),
        }
    }

    fn handle_admin_delete(&self, ctx: &dyn Context, msg: &mut Message) -> Result_ {
        let id = msg.var("id");
        if id.is_empty() {
            return err_bad_request(msg, "Missing document ID");
        }
        match db::delete(ctx, COLLECTION, id) {
            Ok(()) => json_respond(msg, &serde_json::json!({"deleted": true})),
            Err(e) if e.code == "not_found" => err_not_found(msg, "Document not found"),
            Err(e) => err_internal(msg, &format!("Database error: {e}")),
        }
    }

    fn seed_defaults(&self, ctx: &dyn Context) {
        let count = db::count(ctx, COLLECTION, &[]).unwrap_or(0);
        if count > 0 {
            return;
        }

        let now = chrono::Utc::now().to_rfc3339();
        for (doc_type, title, content) in &[
            ("terms", "Terms of Service", "<p>These are the default terms of service. Please update them in the admin panel.</p>"),
            ("privacy", "Privacy Policy", "<p>This is the default privacy policy. Please update it in the admin panel.</p>"),
        ] {
            let mut data = HashMap::new();
            data.insert("doc_type".to_string(), serde_json::Value::String(doc_type.to_string()));
            data.insert("title".to_string(), serde_json::Value::String(title.to_string()));
            data.insert("content".to_string(), serde_json::Value::String(content.to_string()));
            data.insert("status".to_string(), serde_json::Value::String("published".to_string()));
            data.insert("version".to_string(), serde_json::Value::Number(1.into()));
            data.insert("created_at".to_string(), serde_json::Value::String(now.clone()));
            data.insert("updated_at".to_string(), serde_json::Value::String(now.clone()));
            data.insert("published_at".to_string(), serde_json::Value::String(now.clone()));
            data.insert("created_by".to_string(), serde_json::Value::String("system".to_string()));
            if let Err(e) = db::create(ctx, COLLECTION, data) {
                tracing::warn!("Failed to seed default legal page '{doc_type}': {e}");
            }
        }
    }
}

/// Remove dangerous HTML tags and their contents from admin-authored content.
/// Strips `<script>`, `<iframe>`, `<object>`, and `<embed>` tags to prevent stored XSS.
fn sanitize_html(input: &str) -> String {
    let mut s = input.to_string();
    for tag in &["script", "iframe", "object", "embed"] {
        loop {
            let lower = s.to_lowercase();
            let open = format!("<{}", tag);
            if let Some(start) = lower.find(&open) {
                let close = format!("</{}>", tag);
                let end = if let Some(rel_end) = lower[start..].find(&close) {
                    start + rel_end + close.len()
                } else if let Some(gt) = s[start..].find('>') {
                    start + gt + 1
                } else {
                    s.len()
                };
                s = format!("{}{}", &s[..start], &s[end..]);
            } else {
                break;
            }
        }
    }
    s
}

impl Block for LegalPagesBlock {
    fn info(&self) -> BlockInfo {
        BlockInfo {
            name: "@solobase/legalpages".to_string(),
            version: "1.0.0".to_string(),
            interface: "http.handler".to_string(),
            summary: "Legal pages management with versioning and publishing".to_string(),
            instance_mode: InstanceMode::Singleton,
            allowed_modes: vec![InstanceMode::Singleton],
            admin_ui: Some(AdminUIInfo {
                path: "/ext/legalpages/admin".to_string(),
                icon: "Scale".to_string(),
                title: "Legal Pages".to_string(),
            }),
            runtime: wafer_run::types::BlockRuntime::Native,
            requires: Vec::new(),
        }
    }

    fn handle(&self, ctx: &dyn Context, msg: &mut Message) -> Result_ {
        let action = msg.action();
        let path = msg.path();

        match (action, path) {
            // Public endpoints
            ("retrieve", "/ext/legalpages/terms") => self.handle_get_public(ctx, msg, "terms"),
            ("retrieve", "/ext/legalpages/privacy") => self.handle_get_public(ctx, msg, "privacy"),
            // Admin UI
            ("retrieve", "/ext/legalpages/admin") => self.handle_admin_ui(msg),
            // Admin API
            ("retrieve", "/admin/legalpages/documents") => self.handle_admin_list(ctx, msg),
            ("retrieve", _) if path.starts_with("/admin/legalpages/documents/") => self.handle_admin_get(ctx, msg),
            ("create", "/admin/legalpages/documents") => self.handle_admin_create(ctx, msg),
            ("update", _) if path.starts_with("/admin/legalpages/documents/") && path.ends_with("/publish") => {
                self.handle_admin_publish(ctx, msg)
            }
            ("update", _) if path.starts_with("/admin/legalpages/documents/") => self.handle_admin_update(ctx, msg),
            ("delete", _) if path.starts_with("/admin/legalpages/documents/") => self.handle_admin_delete(ctx, msg),
            // ext API aliases (same as admin, but routed through admin-pipe)
            ("retrieve", "/ext/legalpages/documents") => self.handle_admin_list(ctx, msg),
            ("create", "/ext/legalpages/documents") => self.handle_admin_create(ctx, msg),
            _ => err_not_found(msg, "not found"),
        }
    }

    fn lifecycle(&self, ctx: &dyn Context, event: LifecycleEvent) -> std::result::Result<(), WaferError> {
        if matches!(event.event_type, LifecycleType::Init) {
            self.seed_defaults(ctx);
        }
        Ok(())
    }
}

const ADMIN_HTML: &str = r#"<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Legal Pages Admin</title>
<style>
body{font-family:system-ui,sans-serif;margin:0;padding:20px;background:#f5f5f5}
.container{max-width:900px;margin:0 auto}
h1{color:#333}
.card{background:white;border-radius:8px;padding:20px;margin:10px 0;box-shadow:0 1px 3px rgba(0,0,0,0.1)}
.btn{padding:8px 16px;border:none;border-radius:4px;cursor:pointer;font-size:14px;margin:4px}
.btn-primary{background:#6366f1;color:white}
.btn-success{background:#22c55e;color:white}
.btn-danger{background:#ef4444;color:white}
.badge{display:inline-block;padding:2px 8px;border-radius:12px;font-size:12px}
.badge-published{background:#dcfce7;color:#166534}
.badge-draft{background:#fef3c7;color:#92400e}
.badge-archived{background:#e5e7eb;color:#374151}
table{width:100%;border-collapse:collapse}
th,td{padding:10px;text-align:left;border-bottom:1px solid #eee}
</style>
</head>
<body>
<div class="container">
<h1>Legal Pages</h1>
<p>Manage your Terms of Service and Privacy Policy documents.</p>
<div class="card">
<p>Use the admin API endpoints to manage documents:</p>
<ul>
<li><code>GET /admin/legalpages/documents</code> - List all documents</li>
<li><code>POST /admin/legalpages/documents</code> - Create a document</li>
<li><code>PATCH /admin/legalpages/documents/:id</code> - Update a document</li>
<li><code>PATCH /admin/legalpages/documents/:id/publish</code> - Publish a document</li>
<li><code>DELETE /admin/legalpages/documents/:id</code> - Delete a document</li>
</ul>
</div>
</div>
</body>
</html>"#;

File: crates/solobase-cloudflare/worker/blocks-ts/legalpages.ts
// TypeScript-native legalpages block handler (for testing without WASM).
// Calls host.callBlock() for database operations — same pattern as WASM blocks.

import type { Message, BlockResult } from '../types';
import type { RuntimeHost } from '../host';
import { metaGet } from '../convert';

const COLLECTION = 'block_legalpages_legal_documents';

export async function handle(msg: Message, host: RuntimeHost): Promise<BlockResult> {
  const action = metaGet(msg.meta, 'req.action') ?? '';
  const path = metaGet(msg.meta, 'req.resource') ?? '';

  switch (true) {
    // Public endpoints
    case action === 'retrieve' && path === '/b/legalpages/terms':
      return handleGetPublic(msg, host, 'terms');
    case action === 'retrieve' && path === '/b/legalpages/privacy':
      return handleGetPublic(msg, host, 'privacy');

    // Admin API
    case action === 'retrieve' && path === '/admin/legalpages/documents':
      return handleAdminList(msg, host);
    case action === 'retrieve' && path.startsWith('/admin/legalpages/documents/'):
      return handleAdminGet(msg, host);
    case action === 'create' && path === '/admin/legalpages/documents':
      return handleAdminCreate(msg, host);
    case action === 'update' && path.startsWith('/admin/legalpages/documents/') && path.endsWith('/publish'):
      return handleAdminPublish(msg, host);
    case action === 'update' && path.startsWith('/admin/legalpages/documents/'):
      return handleAdminUpdate(msg, host);
    case action === 'delete' && path.startsWith('/admin/legalpages/documents/'):
      return handleAdminDelete(msg, host);

    // Ext API aliases (same as admin, routed through admin-pipe)
    case action === 'retrieve' && path === '/b/legalpages/documents':
      return handleAdminList(msg, host);
    case action === 'create' && path === '/b/legalpages/documents':
      return handleAdminCreate(msg, host);

    default:
      return errResult('not-found', 'not found');
  }
}

// ---------------------------------------------------------------------------
// Public endpoint: serve published legal document as HTML
// ---------------------------------------------------------------------------

async function handleGetPublic(msg: Message, host: RuntimeHost, docType: string): Promise<BlockResult> {
  const result = await callService(host, 'wafer-run/database', 'database.list', {
    collection: COLLECTION,
    filters: [
      { field: 'doc_type', operator: 'eq', value: docType },
      { field: 'status', operator: 'eq', value: 'published' },
    ],
    sort: [{ field: 'version', desc: true }],
    limit: 1,
    offset: 0,
  });

  const records = result?.records ?? [];
  if (records.length === 0) {
    const title = docType === 'terms' ? 'Terms of Service' : 'Privacy Policy';
    const html = `<html><body><h1>${title}</h1><p>No ${docType} document has been published yet.</p></body></html>`;
    return htmlRespond(msg, html);
  }

  const record = records[0];
  const rawContent = record.data?.content ?? '';
  const content = sanitizeHtml(rawContent);
  const title = escapeHtml(record.data?.title ?? docType);

  const html = `<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>${title}</title>
<style>body{font-family:system-ui,sans-serif;max-width:800px;margin:40px auto;padding:0 20px;line-height:1.6;color:#333}h1{color:#111}</style>
</head><body><h1>${title}</h1><div>${content}</div></body></html>`;

  return htmlRespond(msg, html);
}

// ---------------------------------------------------------------------------
// Admin: list documents (paginated, optional type filter)
// ---------------------------------------------------------------------------

async function handleAdminList(msg: Message, host: RuntimeHost): Promise<BlockResult> {
  const { limit, offset } = paginationParams(msg, 20);
  const docType = metaGet(msg.meta, 'req.query.type') ?? '';

  const filters: { field: string; operator: string; value: string }[] = [];
  if (docType) {
    filters.push({ field: 'doc_type', operator: 'eq', value: docType });
  }

  const result = await callService(host, 'wafer-run/database', 'database.list', {
    collection: COLLECTION,
    filters,
    sort: [{ field: 'updated_at', desc: true }],
    limit,
    offset,
  });

  if (!result) return errResult('internal', 'Database error');
  return jsonRespond(msg, result);
}

// ---------------------------------------------------------------------------
// Admin: get single document
// ---------------------------------------------------------------------------

async function handleAdminGet(msg: Message, host: RuntimeHost): Promise<BlockResult> {
  const id = extractDocId(msg);
  if (!id) return errResult('invalid-argument', 'Missing document ID');

  const record = await dbGet(host, COLLECTION, id);
  if (!record) return errResult('not-found', 'Document not found');
  return jsonRespond(msg, record);
}

// ---------------------------------------------------------------------------
// Admin: create document
// ---------------------------------------------------------------------------

async function handleAdminCreate(msg: Message, host: RuntimeHost): Promise<BlockResult> {
  const body = parseBody<{ doc_type: string; title: string; content: string }>(msg);
  if (!body || !body.doc_type || !body.title || body.content === undefined) {
    return errResult('invalid-argument', 'Invalid body: doc_type, title, and content are required');
  }

  const userId = metaGet(msg.meta, 'auth.user_id') ?? '';
  const now = new Date().toISOString();

  const record = await dbCreate(host, COLLECTION, {
    doc_type: body.doc_type,
    title: body.title,
    content: body.content,
    status: 'draft',
    version: 1,
    created_by: userId,
    created_at: now,
    updated_at: now,
  });

  if (!record) return errResult('internal', 'Database error');
  return jsonRespond(msg, record);
}

// ---------------------------------------------------------------------------
// Admin: update document
// ---------------------------------------------------------------------------

async function handleAdminUpdate(msg: Message, host: RuntimeHost): Promise<BlockResult> {
  const id = extractDocId(msg);
  if (!id) return errResult('invalid-argument', 'Missing document ID');

  const body = parseBody<Record<string, unknown>>(msg);
  if (!body) return errResult('invalid-argument', 'Invalid body');

  const data: Record<string, unknown> = { ...body, updated_at: new Date().toISOString() };

  const record = await dbUpdate(host, COLLECTION, id, data);
  if (!record) return errResult('not-found', 'Document not found');
  return jsonRespond(msg, record);
}

// ---------------------------------------------------------------------------
// Admin: publish document
// ---------------------------------------------------------------------------

async function handleAdminPublish(msg: Message, host: RuntimeHost): Promise<BlockResult> {
  const path = metaGet(msg.meta, 'req.resource') ?? '';
  // Extract ID from /admin/legalpages/documents/{id}/publish
  const stripped = path.replace('/admin/legalpages/documents/', '').replace('/publish', '');
  const id = stripped;
  if (!id) return errResult('invalid-argument', 'Missing document ID');

  // Get current document
  const doc = await dbGet(host, COLLECTION, id);
  if (!doc) return errResult('not-found', 'Document not found');

  const docType = doc.data?.doc_type ?? '';

  // Unpublish other documents of same type
  const existing = await callService(host, 'wafer-run/database', 'database.list', {
    collection: COLLECTION,
    filters: [
      { field: 'doc_type', operator: 'eq', value: docType },
      { field: 'status', operator: 'eq', value: 'published' },
    ],
    sort: [],
    limit: 100,
    offset: 0,
  });

  if (existing?.records) {
    for (const r of existing.records) {
      await dbUpdate(host, COLLECTION, r.id, { status: 'archived' });
    }
  }

  // Publish this one
  const now = new Date().toISOString();
  const record = await dbUpdate(host, COLLECTION, id, {
    status: 'published',
    published_at: now,
    updated_at: now,
  });

  if (!record) return errResult('internal', 'Database error');
  return jsonRespond(msg, record);
}

// ---------------------------------------------------------------------------
// Admin: delete document
// ---------------------------------------------------------------------------

async function handleAdminDelete(msg: Message, host: RuntimeHost): Promise<BlockResult> {
  const id = extractDocId(msg);
  if (!id) return errResult('invalid-argument', 'Missing document ID');

  const result = await callService(host, 'wafer-run/database', 'database.delete', {
    collection: COLLECTION,
    id,
  });

  if (!result) return errResult('not-found', 'Document not found');
  return jsonRespond(msg, { deleted: true });
}

// ---------------------------------------------------------------------------
// HTML sanitization (port of Rust sanitize_html)
// ---------------------------------------------------------------------------

const DANGEROUS_TAGS = [
  'script', 'iframe', 'object', 'embed', 'style', 'form',
  'input', 'textarea', 'select', 'button', 'meta', 'link',
  'base', 'svg', 'math', 'applet',
];

function sanitizeHtml(input: string): string {
  let s = input;

  // Strip dangerous tags and their contents
  for (const tag of DANGEROUS_TAGS) {
    let iterations = 0;
    while (iterations++ < 100) {
      const lower = s.toLowerCase();
      const open = `<${tag}`;
      const startIdx = lower.indexOf(open);
      if (startIdx === -1) break;

      const close = `</${tag}>`;
      const closeIdx = lower.indexOf(close, startIdx);
      let endIdx: number;
      if (closeIdx !== -1) {
        endIdx = closeIdx + close.length;
      } else {
        const gtIdx = s.indexOf('>', startIdx);
        endIdx = gtIdx !== -1 ? gtIdx + 1 : s.length;
      }
      s = s.substring(0, startIdx) + s.substring(endIdx);
    }
  }

  // Strip event handler attributes and dangerous URIs from remaining tags
  let result = '';
  let i = 0;
  while (i < s.length) {
    if (s[i] === '<') {
      const endIdx = s.indexOf('>', i);
      if (endIdx !== -1) {
        const tagContent = s.substring(i, endIdx + 1);
        result += removeDangerousAttrs(tagContent);
        i = endIdx + 1;
      } else {
        result += s[i];
        i++;
      }
    } else {
      result += s[i];
      i++;
    }
  }

  return result;
}

function removeDangerousAttrs(tag: string): string {
  // Find end of tag name
  let nameEnd = 1;
  while (nameEnd < tag.length && tag[nameEnd] !== ' ' && tag[nameEnd] !== '>' && tag[nameEnd] !== '/') {
    nameEnd++;
  }

  if (nameEnd >= tag.length || tag[nameEnd] === '>') {
    return tag;
  }

  const tagName = tag.substring(0, nameEnd);
  const rest = tag.substring(nameEnd);
  let result = tagName;
  let pos = 0;

  while (pos < rest.length) {
    // Skip whitespace
    while (pos < rest.length && (rest[pos] === ' ' || rest[pos] === '\t' || rest[pos] === '\n')) {
      result += rest[pos];
      pos++;
    }
    if (pos >= rest.length) break;
    if (rest[pos] === '>' || (rest[pos] === '/' && pos + 1 < rest.length && rest[pos + 1] === '>')) {
      result += rest.substring(pos);
      break;
    }

    // Read attribute name
    const attrStart = pos;
    while (pos < rest.length && rest[pos] !== '=' && rest[pos] !== '>' && rest[pos] !== ' ' && rest[pos] !== '\t') {
      pos++;
    }
    const attrName = rest.substring(attrStart, pos).toLowerCase();

    const isDangerous = attrName.startsWith('on') || attrName === 'srcdoc' || attrName === 'formaction';

    // Read = and value if present
    let attrEnd = pos;
    if (pos < rest.length && rest[pos] === '=') {
      pos++; // skip =
      if (pos < rest.length && (rest[pos] === '"' || rest[pos] === "'")) {
        const quote = rest[pos];
        pos++;
        while (pos < rest.length && rest[pos] !== quote) pos++;
        if (pos < rest.length) pos++; // skip closing quote
      } else {
        while (pos < rest.length && rest[pos] !== ' ' && rest[pos] !== '>') pos++;
      }
      attrEnd = pos;

      const attrValue = rest.substring(attrStart, attrEnd).toLowerCase();
      const hasDangerousUri = attrValue.includes('javascript:') || attrValue.includes('data:text/html') || attrValue.includes('vbscript:');

      if (!isDangerous && !hasDangerousUri) {
        result += rest.substring(attrStart, attrEnd);
      }
    } else if (!isDangerous) {
      result += rest.substring(attrStart, attrEnd);
    }
  }

  return result;
}

// ---------------------------------------------------------------------------
// Service call helpers
// ---------------------------------------------------------------------------

async function callService(host: RuntimeHost, block: string, kind: string, data: unknown): Promise<any> {
  const result = await host.callBlock(block, {
    kind, data: new TextEncoder().encode(JSON.stringify(data)), meta: [],
  });
  if (result.action !== 'respond' || !result.response) return null;
  try { return JSON.parse(new TextDecoder().decode(result.response.data)); } catch { return null; }
}

async function dbGet(host: RuntimeHost, collection: string, id: string): Promise<{ id: string; data: Record<string, any> } | null> {
  return callService(host, 'wafer-run/database', 'database.get', { collection, id });
}

async function dbCreate(host: RuntimeHost, collection: string, data: Record<string, unknown>): Promise<{ id: string; data: Record<string, any> } | null> {
  return callService(host, 'wafer-run/database', 'database.create', { collection, data });
}

async function dbUpdate(host: RuntimeHost, collection: string, id: string, data: Record<string, unknown>): Promise<{ id: string; data: Record<string, any> } | null> {
  return callService(host, 'wafer-run/database', 'database.update', { collection, id, data });
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function extractDocId(msg: Message): string {
  const path = metaGet(msg.meta, 'req.resource') ?? '';
  const suffix = path.startsWith('/admin/legalpages/documents/')
    ? path.substring('/admin/legalpages/documents/'.length)
    : path.startsWith('/b/legalpages/documents/')
      ? path.substring('/b/legalpages/documents/'.length)
      : '';
  // Strip trailing /publish or /
  return suffix.split('/')[0] || '';
}

function paginationParams(msg: Message, defaultLimit: number): { limit: number; offset: number } {
  const page = parseInt(metaGet(msg.meta, 'req.query.page') ?? '1', 10) || 1;
  const limit = parseInt(metaGet(msg.meta, 'req.query.page_size') ?? String(defaultLimit), 10) || defaultLimit;
  const offset = (page - 1) * limit;
  return { limit, offset };
}

function parseBody<T>(msg: Message): T | null {
  try { return JSON.parse(new TextDecoder().decode(msg.data)) as T; } catch { return null; }
}

function escapeHtml(s: string): string {
  return s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

function jsonRespond(msg: Message, data: unknown, status?: number): BlockResult {
  const meta = [{ key: 'resp.content_type', value: 'application/json' }];
  if (status) meta.push({ key: 'resp.status', value: String(status) });
  return {
    action: 'respond',
    response: { data: new TextEncoder().encode(JSON.stringify(data)), meta },
    message: msg,
  };
}

function htmlRespond(msg: Message, html: string): BlockResult {
  return {
    action: 'respond',
    response: {
      data: new TextEncoder().encode(html),
      meta: [{ key: 'resp.content_type', value: 'text/html; charset=utf-8' }],
    },
    message: msg,
  };
}

function errResult(code: string, message: string): BlockResult {
  return { action: 'error', error: { code: code as any, message, meta: [] } };
}
